### ISSUE TITLE
Pythran much slower than Numpy to generate random numbers

### ISSUE THREAD

**Issue by paugier** [serge-sans-paille/pythran#1111]
### Pythran slower than Numpy...

We have a big performance issue with random number generation:

```
# tmp_rand.py
import numpy as np
# pythran export rand2d(int)
def rand2d(n):
    return np.random.rand(n, n)
```
compiled with `pythran -march=native -DUSE_XSIMD tmp_rand.py -o tmp_rand_pythran.so`

With Numpy:
```
In [1]: from tmp_rand import rand2d as rand2d_py                                  
In [2]: for n in [100, 1000, 2000, 4000, 10000]: 
   ...:     %timeit rand2d_py(n) 
   ...:                                                                           
107 µs ± 32.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
10.5 ms ± 3.31 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
42.1 ms ± 6.91 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
217 ms ± 204 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.35 s ± 1.86 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

With Pythran:
```
In [1]: from tmp_rand_pythran import rand2d                                       
In [2]: for n in [100, 1000, 2000, 4000, 10000]: 
   ...:     %timeit rand2d(n) 
   ...:                                                                           
942 µs ± 322 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
94.2 ms ± 68.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
378 ms ± 2.75 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.56 s ± 422 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
9.72 s ± 875 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Oups, Pythran is approximately 10 times slower than Numpy to generate random numbers...

Since many benchmarks involve random number generation, Pythran gets really bad results for these benchmarks because of this random generation issue!  I know that random number generation performance is not be so important in most codes, but communication matters...So silly micro-benchmarks matter :slightly_smiling_face: 

Note that it seems to be a regression since Numpy and Pythran performance were similar some months ago (#759 ).

### Numpy much slower than other solutions for random number generation

Note that random number generation is also known to be quite slow with Numpy (see for example https://software.intel.com/en-us/blogs/2016/06/15/faster-random-number-generation-in-intel-distribution-for-python and https://github.com/numpy/numpy/issues/6967).

Julia random generation is very efficient (and Julia micro-benchmarks are full of random number generation...). https://github.com/fluiddyn/BenchmarksPythonJuliaAndCo/tree/master/JuMicroBenchmarks shows that Numpy is 7.5 slower than Julia (so Pythran is ~70 slower than Julia) for this!

randomgen (https://github.com/bashtage/randomgen) is a serious Python package (code in Cython) for quite fast random number generation.

For random number generation in benchmarks, it would make sens to use it so I think it would make sens to support a tiny part of the API of randomgen in Pythran. Just this would be sufficient in many cases for benchmarks:

```
from randomgen import RandomGenerator
# Default basic PRNG is Xoroshiro128
rnd = RandomGenerator()
rnd.random_uintegers(1000000, bits=32)
rnd.rand(1000, 1000)
```

**Comment by serge-sans-paille** [comment]
pythran currently relies on libstdc++ for PRNG. It would make sens to use exactly the same PRNG as Numpy, that would help for reproducibility, I'll check that.

**Comment by paugier** [comment]
Indication: the problem seems to come from clang... With gcc, Pythran performances are very similar than Numpy. So no "regression" :slightly_smiling_face: 

```
In [1]: from tmp_rand_pythran import rand2d                                     
In [2]: for n in [100, 1000, 2000, 4000, 10000]:  
   ...:     %timeit rand2d(n)  
   ...:                                                                         
75.9 µs ± 20 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
7.53 ms ± 3.25 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
30.2 ms ± 25 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
167 ms ± 28.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.03 s ± 448 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Anyway, there is a problem with clang...

**Comment by serge-sans-paille** [comment]
@paugier can you confirm that #1118 is okay to you, speed-wise? I get good results on my side.

**Comment by paugier** [comment]
Thank you!

The speed up compared to what we had before this commit is impressive! It is now a little bit faster than numpy.random.rand. For the record, Pythran rand is still 4.4 times slower than Julia and 1.6 times slower than the default RG used by randomgen. So it is not super fast but it is fast enough not to be a real problem for Julia's benchmarks.

Here with Python using randomgen:
```
                         | python/julia | pythran/julia | python/pythran |
matrix_statistics_ones   |       6.56   |      0.92     |      7.15      |
matrix_multiply_ones     |       1.01   |      0.99     |      1.03      |
random                   |       2.77   |      4.44     |      0.62      |
parse_integers_rand      |       4.44   |      3.66     |      1.21      |
matrix_statistics_rand   |       6.50   |      1.30     |      4.99      |
matrix_multiply_rand     |       1.27   |      1.49     |      0.85      |
```

I see that you switched to PCG64 (?). So I guess you think it is OK to break the Numpy tests on RNG? I understood that some Numpy developers were very conservative on RNG, which is why Numpy still uses the same random engine and maybe also why https://github.com/bashtage/randomgen has not been included in Numpy.

I have no problem with this choice because (i) I don't really care about the quality of the RNG and (ii) PCG seems to be good but I hope it won't be a problem in future. I guess if people are careful about this, it can be important.

Another more conservative choice would be to support:
```
from randomgen import RandomGenerator, PCG64
rnd = RandomGenerator(PCG64())
rnd.random_uintegers(1000000, bits=32)
rnd.rand(1000, 1000)
```
which is I think equivalent to what you've done and to still use std::mt19937 for the numpy.random API. But this does not solve the issue of std::mt19937 being so slow with clang...

Two small questions showing my bad understanding of C++... Do you think xsimd could help for  ̀discard`? `operator()()` is going to be automatically inlined?

**Comment by serge-sans-paille** [comment]
> I see that you switched to PCG64 (?).
yes, because

1. It's ultra simple to implement (no large library to redistribute)
2. The license is OK
3. From all the sources I gathered, it behaves better than MT

> So I guess you think it is OK to break the Numpy tests on RNG?

As long as the above constraints are respected, yes. But see below first, a vectorized version of PCG would be a great option too. I did check the two points (i) and (ii) you raised before making this choice :-)

> Do you think xsimd could help for discard
for pcg, discard is super fast, as we only need to perform part of the computation
for PCG again, a vectorized version looks feasible (cc @wolfv)
> operator()() is going to be automatically inlined?
You cannot enforce it, unless you mark it as ``always inline``, but that's not the case in the stt (rightfully)
